### PR TITLE
🤖 Code analyzers

### PR DESCRIPTION
**Pull request by heytherewill**
### What ?
This is an initial attempt at enforcing code style using analyzers. This PR does not include CodeFixes because I wanna discuss things before implementing them.

### Reviewing

This PR adds ~~two~~ three analyzers:

- One for components, which ensures they are readonly structs
- One for messages, which ensures they are readonly structs
- One for systems, which ensures the relevant systems are annotated with the correct attributes.

~~The former issues warnings while the later issues compilation errors. I believe we can be even more aggressive and just outright have errors for everything, but I'll leave it up to the project owners.~~. Now everything issues errors except for when you annotate a non-Messager system with the message attribute OR annotate a non-Reactive system with the Watch attribute. These warn you that the attribute will be ignored but don't error out.

Don't check just the code, but also make sure the identifiers are appropriate, the message have adequate grammar and point edge cases which might break the current behavior.

### Next
My other Analyzer ideas are:

~~- Ensuring messages follow the same rules as components (which makes me wonder if there's a reason why IMessage is not an IComponent).~~ ✅
~~- Ensuring the Watch and Messager attributes are only applied to the correct systems~~ ✅
- Ensuring the types provided to the Watch/Messager attributes are valid Components/Messages, respectively

Which I'm more than happy to discuss and implement :)

### CODE REVIEW

File: src/Bang.Analyzers/DiagnosticIds.cs
Comment (isadorasophia) [inline src/Bang.Analyzers/DiagnosticIds.cs:9]: I love this error id 😭

**Comment by isadorasophia** [comment]
Regarding some of the points in the description:
>Ensuring messages follow the same rules as components (which makes me wonder if there's a reason why IMessage is not an IComponent).

This is because messages are not *really* stored in the entity and are extremely cheap to turn on/off (they are just a boolean, really), so that's the main reason I treat them differentely.

Otherwise I love all these ideas! :)

File: src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj
Comment (isadorasophia) [inline src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj]: Yess!! Tests!!! :D
One question though: why did you add xunit and Microsoft.NET.Sdk? I was using just `Microsoft.NET.Test.Sdk` for Gum and it was working all right.

File: src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj
Comment (isadorasophia) [inline src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj]: Why did you add `coverlet.collector` here?

File: src/Bang.Analyzers.Tests/BangAnalyzerVerifier.cs
Comment (isadorasophia) [inline src/Bang.Analyzers.Tests/BangAnalyzerVerifier.cs:36]: Are they adding support in net8.0? I wonder if we should just switch .net 8 pretty soon...

**Review by isadorasophia — CHANGES_REQUESTED** [review]
Thank you for adding those!!! :) (Sorry it took a while, I need to fix my GitHub notifications)

File: src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj
Comment (heytherewill) [reply src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj]: TBH this is just the default Rider template for unit tests and xUnit is my favourite test framework 😂 

I'm happy to use whatever you are more comfortable with, so I can drop xUnit if you so wish

File: src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj
Comment (heytherewill) [reply src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj]: Default Rider test template adds coverlet. It's a nice plus to have code coverage, should we ever wanna do something with it (mostly bragging right methinks)

File: src/Bang.Analyzers.Tests/BangAnalyzerVerifier.cs
Comment (heytherewill) [reply src/Bang.Analyzers.Tests/BangAnalyzerVerifier.cs:36]: I'm not sure about the roadmap, but adapting this Hack to work with net8.0 should be easy enough

File: src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj
Comment (isadorasophia) [reply src/Bang.Analyzers.Tests/Bang.Analyzers.Tests.csproj]: Yup, that would be great! I want to keep all the .csproj as clean as possible hehe

**Comment by heytherewill** [comment]
@isadorasophia Addressed everything. Checking commit by commit from https://github.com/isadorasophia/bang/pull/2/commits/f265a5e3d1d557e330a99d4d74373e92a1c7098f onwards is probably easier than looking at everything :) 

This PR is getting quite large, so since you do approve of the overall structure I suggest merging it and then we can start two other PRs:

- One for running the tests on CI to ensure no regressions happen
- One for actually including the Analyzers in the Bang nuget package

**Review by isadorasophia — APPROVED** [review]
Feel free to merge it :) thank you for all the changes!! (sorry it took a while, busy month T_T)

File: src/Bang.Analyzers/AnalyzerReleases.Shipped.md
Comment (isadorasophia) [inline src/Bang.Analyzers/AnalyzerReleases.Shipped.md:1]: Are these .md files generated?

Comment (heytherewill) [reply src/Bang.Analyzers/AnalyzerReleases.Shipped.md:1]: They are to be manually updated (😭) once we actually ship it to reflect what is and what is not included in the released NuGets

**Comment by heytherewill** [comment]
Thanks for the approval and not need to apologize but... I don't have merge permissions 😅

**Comment by isadorasophia** [comment]
I thought it would be okay after my approval! :O Well, I'll merge it! hehe

**Pull request merged**
